File: src/plugins/meson-templates/resources/src/application-gtk4.c
{{include "license.c"}}

#include "config.h"
#include <glib/gi18n.h>

#include "{{prefix}}-application.h"
#include "{{prefix}}-window.h"

struct _{{PreFix}}Application
{
{{if is_adwaita}}
	AdwApplication parent_instance;
{{else}}
	GtkApplication parent_instance;
{{end}}
};

G_DEFINE_FINAL_TYPE ({{PreFix}}Application, {{prefix_}}_application, {{if is_adwaita}}ADW_TYPE_APPLICATION{{else}}GTK_TYPE_APPLICATION{{end}})

{{PreFix}}Application *
{{prefix_}}_application_new (const char        *application_id,
{{spaces}}                  GApplicationFlags  flags)
{
	g_return_val_if_fail (application_id != NULL, NULL);

	return g_object_new ({{PREFIX}}_TYPE_APPLICATION,
	                     "application-id", application_id,
	                     "flags", flags,
	                     "resource-base-path", "{{appid_path}}",
	                     NULL);
}

static void
{{prefix_}}_application_activate (GApplication *app)
{
	GtkWindow *window;

	g_assert ({{PREFIX}}_IS_APPLICATION (app));

	window = gtk_application_get_active_window (GTK_APPLICATION (app));

	if (window == NULL)
		window = g_object_new ({{PREFIX}}_TYPE_WINDOW,
		                       "application", app,
		                       NULL);

	gtk_window_present (window);
}

static void
{{prefix_}}_application_class_init ({{PreFix}}ApplicationClass *klass)
{
	GApplicationClass *app_class = G_APPLICATION_CLASS (klass);

	app_class->activate = {{prefix_}}_application_activate;
}

static void
{{prefix_}}_application_about_action (GSimpleAction *action,
{{spaces}}                           GVariant      *parameter,
{{spaces}}                           gpointer       user_data)
{
{{if is_adwaita}}
	static const char *developers[] = {"{{author}}", NULL};
{{else}}
	static const char *authors[] = {"{{author}}", NULL};
{{end}}
	{{PreFix}}Application *self = user_data;
	GtkWindow *window = NULL;

	g_assert ({{PREFIX}}_IS_APPLICATION (self));

	window = gtk_application_get_active_window (GTK_APPLICATION (self));

{{if is_adwaita}}
	adw_show_about_dialog (GTK_WIDGET (window),
	                       "application-name", "{{name}}",
	                       "application-icon", "{{appid}}",
	                       "developer-name", "{{author}}",
	                       "translator-credits", _("translator-credits"),
	                       "version", "{{project_version}}",
	                       "developers", developers,
	                       "copyright", "© {{year}} {{author}}",
	                       NULL);
{{else}}
	gtk_show_about_dialog (window,
	                       "program-name", "{{name}}",
	                       "logo-icon-name", "{{appid}}",
	                       "authors", authors,
	                       "translator-credits", _("translator-credits"),
	                       "version", "{{project_version}}",
	                       "copyright", "© {{year}} {{author}}",
	                       NULL);
{{end}}
}

static void
{{prefix_}}_application_quit_action (GSimpleAction *action,
{{spaces}}                          GVariant      *parameter,
{{spaces}}                          gpointer       user_data)
{
	{{PreFix}}Application *self = user_data;

	g_assert ({{PREFIX}}_IS_APPLICATION (self));

	g_application_quit (G_APPLICATION (self));
}

static const GActionEntry app_actions[] = {
	{ "quit", {{prefix_}}_application_quit_action },
	{ "about", {{prefix_}}_application_about_action },
};

static void
{{prefix_}}_application_init ({{PreFix}}Application *self)
{
	g_action_map_add_action_entries (G_ACTION_MAP (self),
	                                 app_actions,
	                                 G_N_ELEMENTS (app_actions),
	                                 self);
	gtk_application_set_accels_for_action (GTK_APPLICATION (self),
	                                       "app.quit",
	                                       (const char *[]) { "<primary>q", NULL });
}

File: src/plugins/meson-templates/resources/src/application-gtk4.rs
{{include "license.rs"}}

use gettextrs::gettext;
{{if is_adwaita}}
use adw::prelude::*;
use adw::subclass::prelude::*;
{{else}}
use gtk::prelude::*;
use gtk::subclass::prelude::*;
{{end}}
use gtk::{gio, glib};

use crate::config::VERSION;
use crate::{{PreFix}}Window;

mod imp {
    use super::*;

    #[derive(Debug, Default)]
    pub struct {{PreFix}}Application {}

    #[glib::object_subclass]
    impl ObjectSubclass for {{PreFix}}Application {
        const NAME: &'static str = "{{PreFix}}Application";
        type Type = super::{{PreFix}}Application;
        type ParentType = {{if is_adwaita}}adw::Application{{else}}gtk::Application{{end}};
    }

    impl ObjectImpl for {{PreFix}}Application {
        fn constructed(&self) {
            self.parent_constructed();
            let obj = self.obj();
            obj.setup_gactions();
            obj.set_accels_for_action("app.quit", &["<primary>q"]);
        }
    }

    impl ApplicationImpl for {{PreFix}}Application {
        // We connect to the activate callback to create a window when the application
        // has been launched. Additionally, this callback notifies us when the user
        // tries to launch a "second instance" of the application. When they try
        // to do that, we'll just present any existing window.
        fn activate(&self) {
            let application = self.obj();
            // Get the current window or create one if necessary
            let window = application.active_window().unwrap_or_else(|| {
                let window = {{PreFix}}Window::new(&*application);
                window.upcast()
            });

            // Ask the window manager/compositor to present the window
            window.present();
        }
    }

    impl GtkApplicationImpl for {{PreFix}}Application {}
    {{if is_adwaita}}
impl AdwApplicationImpl for {{PreFix}}Application {}
{{end}}
}

glib::wrapper! {
    pub struct {{PreFix}}Application(ObjectSubclass<imp::{{PreFix}}Application>)
        @extends gio::Application, gtk::Application, {{if is_adwaita}}adw::Application,{{end}}

        @implements gio::ActionGroup, gio::ActionMap;
}

impl {{PreFix}}Application {
    pub fn new(application_id: &str, flags: &gio::ApplicationFlags) -> Self {
        glib::Object::builder()
            .property("application-id", application_id)
            .property("flags", flags)
            .property("resource-base-path", "{{appid_path}}")
            .build()
    }

    fn setup_gactions(&self) {
        let quit_action = gio::ActionEntry::builder("quit")
            .activate(move |app: &Self, _, _| app.quit())
            .build();
        let about_action = gio::ActionEntry::builder("about")
            .activate(move |app: &Self, _, _| app.show_about())
            .build();
        self.add_action_entries([quit_action, about_action]);
    }

    fn show_about(&self) {
        let window = self.active_window().unwrap();
{{if is_adwaita}}
        let about = adw::AboutDialog::builder()
            .application_name("{{name}}")
            .application_icon("{{appid}}")
            .developer_name("{{author}}")
            .version(VERSION)
            .developers(vec!["{{author}}"])
            // Translators: Replace "translator-credits" with your name/username, and optionally an email or URL.
            .translator_credits(&gettext("translator-credits"))
            .copyright("© {{year}} {{author}}")
            .build();

        about.present(Some(&window));
{{else}}
        let about = gtk::AboutDialog::builder()
            .transient_for(&window)
            .modal(true)
            .program_name("{{name}}")
            .logo_icon_name("{{appid}}")
            .version(VERSION)
            .authors(vec!["{{author}}"])
            // Translators: Replace translator-credits with your name/username, and optionally an email or URL.
            .translator_credits(&gettext("translator-credits"))
            .copyright("© {{year}} {{author}}")
            .build();

        about.present();
{{end}}
    }
}
